core/src/main/java/fr/sncf/osrd/conflicts/Conflicts.kt

```diff
@@ -204,6 +241,119 @@ class IncrementalConflictDetectorImpl(trainRequirements: List<TrainRequirements>
+                        }
+                    if (conflictingRequirements.isNotEmpty()) {
+                        hasConflict = true
+                        val latestEndTime = conflictingRequirements.maxOf { it.endTime }
+                        minDelay = min(minDelay, latestEndTime - routingRequirement.beginTime)
+                    }
+                }
+            }
+        }
+        if (!hasConflict) return globalMinDelay
+        else if (!isFinite(minDelay)) return Double.POSITIVE_INFINITY
+        else {
+            val shiftedSpacingRequirements =
+                spacingRequirements.toMutableList().onEach {
+                    it.beginTime += minDelay
+                    it.endTime += minDelay
+                }
+            val shiftedRoutingRequirements =
+                routingRequirements.toMutableList().onEach { routingRequirement ->
+                    routingRequirement.beginTime += minDelay
+                    routingRequirement.zones.onEach { it.endTime += minDelay }
+                }
+            return minDelayWithoutConflicts(
+                shiftedSpacingRequirements,
+                shiftedRoutingRequirements,
+                globalMinDelay + minDelay
+            )
+        }
+    }
+
+    override fun maxDelayWithoutConflicts(
+        spacingRequirements: List<SpacingRequirement>,
+        routingRequirements: List<RoutingRequirement>
+    ): Double {
+        var maxDelay = Double.POSITIVE_INFINITY
+        for (spacingRequirement in spacingRequirements) {
+            if (spacingZoneRequirements[spacingRequirement.zone!!] != null) {
+                val endTime = spacingRequirement.endTime
+                spacingZoneRequirements[spacingRequirement.zone!!]!!.forEach {
+                    if (endTime <= it.beginTime) maxDelay = min(maxDelay, it.beginTime - endTime)
+                }
+            }
+        }
+        for (routingRequirement in routingRequirements) {
+            for (zoneReq in routingRequirement.zones) {
+                if (routingZoneRequirements[zoneReq.zone!!] != null) {
+                    val endTime = zoneReq.endTime
+                    routingZoneRequirements[zoneReq.zone!!]!!.forEach {
+                        if (endTime <= it.beginTime)
+                            maxDelay = min(maxDelay, it.beginTime - endTime)
+                    }
+                }
+            }
+        }
+        return maxDelay
+    }
+
+    override fun timeOfNextConflict(
+        spacingRequirements: List<SpacingRequirement>,
+        routingRequirements: List<RoutingRequirement>
+    ): Double {
+        var timeOfNextConflict = Double.POSITIVE_INFINITY
+        for (spacingRequirement in spacingRequirements) {
+            if (spacingZoneRequirements[spacingRequirement.zone!!] != null) {
+                val endTime = spacingRequirement.endTime
+                spacingZoneRequirements[spacingRequirement.zone!!]!!.forEach {
```

Comment:
Nitpicky, I don't like `forEach` over a standard `for` loop, it messes with step-by-step debugging